pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/PinotSinkFunction.java

```diff
@@ -117,31 +123,22 @@ public void close()
   @Override
   public void invoke(T value, Context context)
       throws Exception {
-    _segmentWriter.collect(_recordConverter.convertToRow(value));
+    GenericRow row = _recordConverter.convertToRow(value);
+    _pendingRows.add(row);
+    _segmentWriter.collect(row);
```

Comment:
would it be easier if we just checkpoint the segmentWriter?

Reply:
I was thinking about this as well. We would have to modify the segmentWriter so that it can return all pending rows during each checkpoint, but there could be quite a few rows to read and doing this every checkpoint seems inefficient when we could just continually add the pending rows to flink state as they come in?

Reply:
yeah. i misspoke. i meant to use "clearer" not "easier". it is definitely not easier to implement as you mentioned. let's keep it this way. but we should document that the rows are dual copied inside the PinotSinkFunction.